src/Microsoft.OData.Core/UriParser/SemanticAst/ODataPathExtensions.cs

```diff
@@ -42,6 +47,11 @@ public static IEdmTypeReference EdmType(this ODataPath path)
         /// resource that is part of a set.</returns>
         public static IEdmNavigationSource NavigationSource(this ODataPath path)
         {
+            if (path == null)
+            {
+                throw Error.ArgumentNull(nameof(path));
+            }
+
             return path.LastSegment.TranslateWith(new DetermineNavigationSourceTranslator());

```

Comment:
```suggestion
            return path.LastSegment?.TranslateWith(DetermineNavigationSourceTranslator.Instance);
```

Please introduce a singleton instance of `DetermineNavigationSourceTranslator` and use that there to reduce the number of object allocations

Reply:
#Resolved